--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -29,7 +29,7 @@
 #include <binder/Parcel.h>
 #include <cutils/jstring.h>
 #include <sys/types.h>
-#include <sys/limits.h>
+#include <limits.h>
 #include <sys/system_properties.h>
 #include <pwd.h>
 #include <stdio.h>
@@ -47,6 +47,7 @@
 #include <netinet/in.h>
 #include <cutils/properties.h>
 #include <RilSapSocket.h>
+#include <sys/signal.h>

 extern "C" void
 RIL_onRequestComplete(RIL_Token t, RIL_Errno e, void *response, size_t responselen);
